Support complex connector expression filter pushdown in mongodb

krvikash opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description


This PR covers use cases for connector expression Filter pushdown:

- Complex Filter which can not represent by TupleDomain
- `json_extract_scalar`
- `from_iso8601_timestamp`
- `from_iso8601_date`

Covered Logical operator:

- `OR`, `AND`

Covered Comparision Operator:

- `=`, `!=`, `<`, `>`, `>=`, `<=`

Other Operators:

- `IN`
- `IS NULL`
- `IS NOT NULL`

Covered Type:

- `BooleanType`
- `ObjectIdType`
- `CharType`, `VarcharType`
- `TINYINT`, `SMALLINT`, `INT`, `BIGINT`, `DecimalType`
- `TimeType`, `DateType`, `ShortTimestampType`, `ShortTimestampWithTimeZoneType`

TODO: Fix below
```
testJsonExtractScalarPredicatePushdownWithNullAndMissingField
testComplexExpressionPredicatePushdownWithMismatchedDataType
```
TODO: support for `CAST` as follow up PR

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues



<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
(X) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Mongodb
* Fix some things. ({issue}`issuenumber`)
```

Praveen2112 commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoMetadata.java:
```diff
@@ -617,9 +633,37 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(C
+            if (isComplexExpressionPushdown(session)) {
+                ImmutableList.Builder<String> newExpressions = ImmutableList.builder();
+                ImmutableList.Builder<ConnectorExpression> remainingExpressions = ImmutableList.builder();
+                for (ConnectorExpression expression : extractConjuncts(constraint.getExpression())) {
```
Why do we need to extract the Conjuncts ? Why can't we have a single String which could represent or partially represent the filter expression ?

krvikash commented:
Thanks @Praveen2112 for the review. Addressed comments.

Praveen2112 commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/expression/RewriteIsNull.java:
```diff
@@ -0,0 +1,68 @@
+            return Optional.empty();
+        }
+        Object value = rewritten.get().expression();
+        if (arguments.get(0) instanceof Variable) {
```
Can we use `getOnlyElement` here ?

Praveen2112 commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/expression/RewriteComparison.java:
```diff
@@ -0,0 +1,143 @@
+
+        ComparisonOperator operator = ComparisonOperator.forFunctionName(call.getFunctionName());
+        if (captures.get(LEFT) instanceof Variable) {
+            leftValue = "$" + leftValue;
```
Is it possible to move it to a static class ?

Adding a `$` is only for specific operations or ?

krvikash replied:
This part may be not required if I place `$` while rewriting the variable in `RewriteVariable`.

Praveen2112 commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/expression/ExpressionUtils.java:
```diff
@@ -0,0 +1,48 @@
+
+    public static Document toBinary(byte[] value)
+    {
+        return documentOf("$binary", documentOf("base64", Base64.getEncoder().encodeToString(value)).append("subType", "00"));
```
Can we add a comment - on why we need this `subType`

krvikash replied:
As @ebyhr pointed out [here](https://github.com/trinodb/trino/pull/18382#discussion_r1271963843) that We could get incorrect results if subType is not 0. I will remove `varbinary` type filter pushdown support.

krvikash commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/expression/JsonObjectExpressionBuilder.java:
```diff
@@ -0,0 +1,106 @@
+{
+    private final String json;
+    private final List<String> jsonPath;
+    private static int variableCount;
```
variableCount is to generate a new variable name for `$let` operator.

krvikash commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/expression/ExpressionUtils.java:
```diff
@@ -0,0 +1,56 @@
+    {
+        Document convertValue = documentOf("input", value)
+                .append("to", "string")
+                .append("onError", null);
```
If it throws error then we will miss the case where json_extract_scalar is used with ArrayType/Object Type element.

// json_extract_scalar returns NULL for non-scalar types

krvikash commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/expression/MongoExpression.java:
```diff
@@ -0,0 +1,36 @@
+        requireNonNull(expression, "expression is null");
+    }
+
+    public Optional<Document> expressionDocument()
+    {
+        if (expression instanceof Document document) {
+            return Optional.of(document);
+        }
+        return Optional.empty();
```
Since the current implementation of `MongoExpression` can have object as a document or a literal value. So I am returning empty document when `expressionDocument` is called for a literal value object.

krvikash commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/expression/RewriteDateTimeConstant.java:
```diff
@@ -0,0 +1,95 @@
+    @Override
+    public Optional<MongoExpression> rewrite(Constant constant, Captures captures, RewriteContext<MongoExpression> context)
+    {
+        Object value = constant.getValue();
+        Type type = constant.getType();
+        if (value == null) {
+            return Optional.empty();
+        }
```
@Praveen2112 Sorry I could not understand. Could you point out the code URL here?

krvikash commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/expression/RewriteVariable.java:
```diff
@@ -0,0 +1,41 @@
+    public Optional<MongoExpression> rewrite(Variable variable, Captures captures, RewriteContext<MongoExpression> context)
+    {
+        MongoColumnHandle columnHandle = (MongoColumnHandle) context.getAssignment(variable.getName());
+        return Optional.of(new MongoExpression(columnHandle.getQualifiedName()));
```
It will work. I will add a test case for the same.

vlad-lyutenko commented:
LGTM

Praveen2112 commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/expression/RewriteDateTimeConstant.java:
```diff
@@ -0,0 +1,95 @@
+import static java.lang.Math.floorMod;
+import static java.time.ZoneOffset.UTC;
+
+public class RewriteDateTimeConstant
```
Can we split them into multiple rules for each specific data type ?

Praveen2112 replied:
Can we introduce an abstract class for Constant values which handles for null value and we could have different implementation for each type

Praveen2112 commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/expression/RewriteBooleanConstant.java:
```diff
@@ -0,0 +1,44 @@
+    public Optional<MongoExpression> rewrite(Constant constant, Captures captures, RewriteContext<MongoExpression> context)
+    {
+        Boolean value = (Boolean) constant.getValue();
+        return Optional.of(new MongoExpression(value));
```
Do we need to handle for null value ?

krvikash replied:
> Do we need to handle for null value ?

For the case, Where the constant can be null. Like below
```
assertThat(query(session, "SELECT * FROM %1$s WHERE col %2$s %3$s OR another_col IN (NULL, %3$s)".formatted(table.getName(), operator, value)))
                    .isFullyPushedDown();
```

Praveen2112 commented on plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoConnectorTest.java:
```diff
@@ -333,6 +333,100 @@ public void testPredicatePushdown(String value, String operator)
         }
     }
 
+    @Test(dataProvider = "predicatePushdownProvider")
```
Additionally can we enable this flag true for all our query runners and check if it passes the CI

Praveen2112 commented on plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/expression/MongoConnectorExpressionRewriterBuilder.java:
```diff
@@ -46,6 +46,8 @@ public MongoConnectorExpressionRewriterBuilder addDefaultRules(TypeManager typeM
         add(new RewriteIsNull());
         add(new RewriteJsonPath(typeManager.getType(new TypeSignature("JsonPath"))));
         add(new RewriteJsonExtractScalar());
+        add(new RewriteFromIso8601Date());
```
Can we reword the commit message to make it less than 75 ?

Praveen2112 reviewed (commented):
Just skimmed the PR

krvikash commented on plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoConnectorTest.java:
```diff
@@ -351,6 +353,30 @@ public Object[][] predicatePushdownProvider()
+                "(col DECIMAL(34, 0))",
+                ImmutableList.of("decimal '3141592653589793238462643383279502'"))) { // 34 precision decimal value
+            String predicateValue = "decimal '31415926535897932384626433832795028841'"; // 38 precision decimal value
+            assertThat(query("SELECT * FROM " + table.getName() + " WHERE col = " + predicateValue))
```
https://github.com/trinodb/trino/pull/18382#discussion_r1273098925

In this case, I was not able to apply pushdown condition.

This is the Explain of the `"SELECT * FROM " + table.getName() + " WHERE col = " + predicateValue`
```
Trino version: testversion
Fragment 0 [SINGLE]
    Output layout: [col]
    Output partitioning: SINGLE []
    Output[columnNames = [col]]
    │   Layout: [col:decimal(34,0)]
    │   Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
    └─ Values[]
           Layout: [col:decimal(34,0)]
           Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
```

krvikash commented on plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoConnectorTest.java:
```diff
@@ -342,6 +342,7 @@ public Object[][] predicatePushdownProvider()
                 {"integer '3'"},
                 {"bigint '4'"},
                 {"'test'"},
+                {"char 'test'"},
```
Pushed changes for this in https://github.com/trinodb/trino/pull/18382 PR.

krvikash commented:
As suggested by @ebyhr, I will break this single PR into multiple PRs according to https://github.com/trinodb/trino/issues/18513.

krvikash commented:
Thanks @Praveen2112 | @vlad-lyutenko for the review. I have addressed comments.

I have separated [Support complex connector expression filter pushdown in mongodb #18367](https://github.com/trinodb/trino/pull/18367) (Closing this PR) in multiple PRS. PRs needs to be review in the same order as below.

Closing this PR in favor of below:

- [ ] https://github.com/trinodb/trino/pull/18382
- [ ] https://github.com/trinodb/trino/pull/18518
- [ ] https://github.com/trinodb/trino/pull/18519
- [ ] https://github.com/trinodb/trino/pull/18520
- [ ] https://github.com/trinodb/trino/pull/18522
